python_modules/dagster/dagster/_core/execution/plan/execute_step.py

```diff
@@ -905,7 +909,7 @@ def _log_asset_materialization_events_for_asset(
         # auto-converted source assets. This can be removed when source asset observation functions
         # are converted to use ObserveResult.
         if (
-            execution_type == AssetExecutionType.OBSERVATION
+            asset_key in asset_layer.source_assets_by_key
```

Comment:
Why this change? I'm trying to eliminate these sorts of `source_assets_by_key` refs and I think the second clause ensures this will only eval to True when it should.

Reply:
If a @multi_observable_source_asset is used, wrap_source_asset_observe_fn_in_op_compute_fn is not involved. I.e. the expectation is that the framework will yield the observation events, not the body of the op itself.

But without this change, we don't get to the code below in this function that yields the observation events.

So it seems like we only want this if statement to return True in the case where there's a SourceAsset involved.

Reply:
`SYSTEM_METADATA_KEY_SOURCE_ASSET_OBSERVATION` is only attached to the `Output` when it comes from `wrap_source_asset_observe_fn_in_op_compute_fn`, so this statement will never return True for `@multi_observable_source_asset` no matter what the first clause is.

But if we change the first clause to check for a source asset, then this will give the wrong behavior and produce double observations events for source assets that have been converted with `create_external_asset_from_source_asset`.

Reply:
Ahh, got it